src/Core/Cart/CartRuleCalculator.php

```diff
@@ -120,18 +120,22 @@ protected function applyCartRule(CartRuleData $cartRuleData)
             /** @var CartRow|null $cartRowCheapest */
             $cartRowCheapest = null;
             foreach ($this->cartRows as $cartRow) {
-                if ($cartRowCheapest === null
-                    || $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded() > $cartRow->getInitialUnitPrice()
-                                                                                           ->getTaxIncluded()
-                ) {
-                    $cartRowCheapest = $cartRow;
+                $product = $cartRow->getRowData();
+                if ((($cartRule->reduction_exclude_special && !$product['reduction_applies'])
+                    || !$cartRule->reduction_exclude_special)) {
+                    if ($cartRowCheapest === null
+                        || $cartRowCheapest->getInitialUnitPrice()->getTaxIncluded() > $cartRow->getInitialUnitPrice()
```

Comment:
`cartRowCheapest` is always null here :thinking: 
This condition seems useless

Reply:
I only added some extra lines to get the exclude_specials working and fix the taxes bug.
I tested it and it works. So i haven't really looked at the other code since it's behaving like it should :)